tidy3d/components/simulation.py

```diff
@@ -2769,3 +2769,8 @@ def volumetric_structures(self) -> Tuple[Structure]:
         """Generate a tuple of structures wherein any 2D materials are converted to 3D
         volumetric equivalents."""
         return self._volumetric_structures_grid(self.grid)
+
+    @cached_property
+    def allow_gain(self) -> bool:
+        """``True`` if any of the mediums in the simulation allow gain."""
+        return np.any([med.allow_gain for med in self.mediums])
```

Comment:
In AnisotropicMedium, only the value of each component's `allow_gain` matters. AnisotropicMedium's `allow_gain` is dummy in the current implementation: https://github.com/flexcompute/tidy3d/blob/37aca77d86c53ad7bdad0df5fe0deebc848039d7/tidy3d/components/medium.py#L2332-L2335

So if `isinstance(med, AnisotropicMedium)`, now it requires an additional step in looking through its components.

Reply:
I see. I think the class tree of the Medium could use some reorganization, I think we're all aware of that, but somewhat tied up before 3.0. But generally it would seem that having separate abstract classes for an isotropic medium, diagonally anisotropic medium, and fully anisotropic medium, such that any actual medium is a subclass of one and exactly one of these, would be great. Actually I'm not sure if even that would solve the problem because of all the custom/anisotropic possibilities.

Not a huge fan on doing things based on `isinstance(med, AnisotropicMedium)` and I know that @tylerflex isn't either. But I don't see an easy way around this.

By the way what about fully anisotropic medium, no gain is allowed right now right? I think you asked about it at some point and we agreed on that.

Reply:
> By the way what about fully anisotropic medium, no gain is allowed right now right? I think you asked about it at some point and we agreed on that.

I don't think I've been involved in discussions on this. Maybe @dbochkov-flexcompute? Currently gain is allowed: https://github.com/flexcompute/tidy3d/blob/37aca77d86c53ad7bdad0df5fe0deebc848039d7/tidy3d/components/medium.py#L2484-L2498

Reply:
Oh, I think I'm just confused with some other discussion.